processor/src/trace/virtual_bus/circuit/mod.rs

```diff
@@ -11,8 +14,83 @@ pub use prover::prove;
 mod verifier;
 pub use verifier::verify;
 
+use super::multilinear::inner_product;
 use super::sum_check::{FinalOpeningClaim, Proof as SumCheckProof};
 
+/// Defines the number of elements for the partial left/right numerator/denominators of
+/// [`LayerGatesInputs`].
+const NUM_ELEMENTS_PER_GATE_INPUT: usize = 4;
+
+/// Holds the contribution of one main trace row to the input layer's gates inputs.
+struct LayerGatesInputs<E: FieldElement> {
+    pub partial_left_numerator: [E; NUM_ELEMENTS_PER_GATE_INPUT],
+    pub partial_right_numerator: [E; NUM_ELEMENTS_PER_GATE_INPUT],
+    pub partial_left_denominator: [E; NUM_ELEMENTS_PER_GATE_INPUT],
+    pub partial_right_denominator: [E; NUM_ELEMENTS_PER_GATE_INPUT],
+}
+
+impl<E: FieldElement> LayerGatesInputs<E> {
+    pub fn from_main_trace_query(query: &[E], log_up_randomness: &[E]) -> Self {
+        Self {
+            partial_left_numerator: Self::left_numerator(query),
+            partial_right_numerator: Self::right_numerator(query),
+            partial_left_denominator: Self::left_denominator(query, log_up_randomness),
+            partial_right_denominator: Self::right_denominator(query, log_up_randomness),
+        }
+    }
+
+    fn left_numerator(query: &[E]) -> [E; NUM_ELEMENTS_PER_GATE_INPUT] {
+        let f_m = {
+            let mem_selec0 = query[CHIPLETS_OFFSET];
+            let mem_selec1 = query[CHIPLETS_OFFSET + 1];
+            let mem_selec2 = query[CHIPLETS_OFFSET + 2];
+            mem_selec0 * mem_selec1 * (E::ONE - mem_selec2)
+        };
+        let f_rc = {
+            let op_bit_4 = query[DECODER_OP_BITS_OFFSET + 4];
+            let op_bit_5 = query[DECODER_OP_BITS_OFFSET + 5];
+            let op_bit_6 = query[DECODER_OP_BITS_OFFSET + 6];
+
+            (E::ONE - op_bit_4) * (E::ONE - op_bit_5) * op_bit_6
+        };
+
+        [query[M_COL_IDX], f_m, f_m, f_rc]
+    }
+
+    fn right_numerator(query: &[E]) -> [E; NUM_ELEMENTS_PER_GATE_INPUT] {
+        let f_rc = {
+            let op_bit_4 = query[DECODER_OP_BITS_OFFSET + 4];
+            let op_bit_5 = query[DECODER_OP_BITS_OFFSET + 5];
+            let op_bit_6 = query[DECODER_OP_BITS_OFFSET + 6];
+
+            (E::ONE - op_bit_4) * (E::ONE - op_bit_5) * op_bit_6
+        };
+
+        [f_rc, f_rc, f_rc, E::ZERO]
```

Comment:
nit: maybe a note about the padding

Reply:
Good idea; created `padding` variable + comment